sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala

```diff
@@ -210,6 +210,37 @@ case class Sort(
   override def output: Seq[Attribute] = child.output
 }
 
+case class Range(
+    start: Long,
+    end: Long,
+    step: Long,
+    numSlices: Int) extends LeafNode {
+  require(step != 0, "step cannot be 0")
+  val numElements: BigInt = {
+    val safeStart = BigInt(start)
+    val safeEnd = BigInt(end)
+    if ((safeEnd - safeStart) % step == 0 || safeEnd > safeStart ^ step > 0) {
+      (safeEnd - safeStart) / step
+    } else {
+      // the remainder has the same sign with range, could add 1 more
+      (safeEnd - safeStart) / step + 1
+    }
+  }
+
+  val output: Seq[Attribute] =
```

Comment:
I think it would be better to make this an argument to the constructor and have a factory in the companion object that creates one automatically.